internal/allocator/allocator_test.go

```diff
@@ -1226,8 +1231,135 @@ func TestPoolCount(t *testing.T) {
+func TestPoolMetrics(t *testing.T) {
+	alloc := New()
+	if err := alloc.SetPools(map[string]*config.Pool{
+		"test": {
+			AutoAssign: true,
+			CIDR: []*net.IPNet{
+				ipnet("1.2.3.4/30"),
+				ipnet("1000::4/126"),
+			},
+		},
+	}); err != nil {
+		t.Fatalf("SetPools: %s", err)
+	}
+
+	tests := []struct {
+		desc       string
+		svc        string
+		ip         string
+		ports      []Port
+		sharingKey string
+		backendKey string
+		ipsInUse   int
+	}{
+		{
+			desc:     "assign s1",
+			svc:      "s1",
+			ip:       "1.2.3.4",
+			ipsInUse: 1,
+		},
+		{
+			desc:     "assign s2",
+			svc:      "s2",
+			ip:       "1.2.3.5",
+			ipsInUse: 2,
+		},
+		{
+			desc:     "unassign s1",
+			svc:      "s1",
+			ipsInUse: 1,
+		},
+		{
+			desc:     "unassign s2",
+			svc:      "s2",
+			ipsInUse: 0,
+		},
+		{
+			desc:       "assign s1 shared",
+			svc:        "s1",
+			ip:         "1.2.3.4",
+			sharingKey: "key",
+			ports:      ports("tcp/80"),
+			ipsInUse:   1,
+		},
+		{
+			desc:       "assign s2 shared",
+			svc:        "s2",
+			ip:         "1.2.3.4",
+			sharingKey: "key",
+			ports:      ports("tcp/443"),
+			ipsInUse:   1,
+		},
+		{
+			desc:       "assign s3 shared",
+			svc:        "s3",
+			ip:         "1.2.3.4",
+			sharingKey: "key",
+			ports:      ports("tcp/23"),
+			ipsInUse:   1,
+		},
+		{
+			desc:     "unassign s1 shared",
+			svc:      "s1",
+			ports:    ports("tcp/80"),
+			ipsInUse: 1,
+		},
+		{
+			desc:     "unassign s2 shared",
+			svc:      "s2",
+			ports:    ports("tcp/443"),
+			ipsInUse: 1,
+		},
+		{
+			desc:     "unassign s3 shared",
+			svc:      "s3",
+			ports:    ports("tcp/23"),
+			ipsInUse: 0,
+		},
+	}
+
+	for _, test := range tests {
+		if test.ip == "" {
+			alloc.Unassign(test.svc)
+			if !gaugeEqual(stats.poolActive.WithLabelValues("test"), test.ipsInUse) {
+				t.Errorf("stats.poolCapacity invalid. Expected %d", test.ipsInUse)
+			}
+			continue
+		}
+
+		ip := net.ParseIP(test.ip)
+		if ip == nil {
+			t.Fatalf("invalid IP %q in test %q", test.ip, test.desc)
+		}
+		err := alloc.Assign(test.svc, ip, test.ports, test.sharingKey, test.backendKey)
+
+		if err != nil {
+			t.Errorf("%q: Assign(%q, %q): %s", test.desc, test.svc, test.ip, err)
+		}
+		if a := assigned(alloc, test.svc); a != test.ip {
+			t.Errorf("%q: ran Assign(%q, %q), but allocator has recorded allocation of %q", test.desc, test.svc, test.ip, a)
+		}
+		if !gaugeEqual(stats.poolCapacity.WithLabelValues("test"), 8) {
+			t.Errorf("stats.poolCapacity invalid. Expected %d", 8)
+		}
+		if !gaugeEqual(stats.poolActive.WithLabelValues("test"), test.ipsInUse) {
+			t.Errorf("stats.poolCapacity invalid %s. Expected %d", test.svc, test.ipsInUse)
+		}
+	}
+}
+
 // Some helpers
 
+func gaugeEqual(gauge prometheus.Gauge, value int) bool {
+	metric := &dto.Metric{}
+	gauge.Write(metric)
+	v := fmt.Sprintf("value: %d", value)
+	// TODO: improve "value: 88" will match "value: 8"
+	return strings.Contains(proto.MarshalTextString(metric), v)
+}
```

Comment:
Yeah, I share the concern with that `strings.Contains()`. But there is a `testutil` package to test this easily. See here: https://github.com/prometheus/client_golang/issues/486#issuecomment-433345239

Basically, you can just import `github.com/prometheus/client_golang/prometheus/testutil` and do something like this (just tested it locally):

```suggestion
func gaugeEqual(gauge prometheus.Gauge, value float64) bool {
      return ptu.ToFloat64(gauge) == value
}
```
And modify the param and test func to use float64.

This will actually test the number is exactly as what we expect it to be. What do you think?

Reply:
much better. I did not find the prometheus testutil package.

Reply:
Thanks!